metals/src/main/scala/scala/meta/internal/metals/mcp/MetalsMcpServer.scala

```diff
@@ -0,0 +1,631 @@
+          )
+          (result match {
+            case Right(value) =>
+              value.map(content =>
+                new CallToolResult(createContent(content), false)
+              )
+            case Left(error) =>
+              Future.successful(
+                new CallToolResult(createContent(s"Error: $error"), true)
+              )
+          }).toMono
+        } catch {
+          case e: Exception =>
+            Mono.just(
+              new CallToolResult(createContent(s"Error: ${e.getMessage}"), true)
+            )
+        }
+      },
+    )
+  }
+
+  private def createGlobSearchTool(): AsyncToolSpecification = {
+    val schema = """
+      {
+        "type": "object",
+        "properties": {
+          "query": {
+            "type": "string",
+            "description": "Substring of the symbol to search for"
+          },
+          "fileInFocus": {
+            "type": "string",
+            "description": "The current file in focus for context, if empty we will try to detect it"
+          }
+        },
+        "required": ["query"]
+      }
+    """
+    new AsyncToolSpecification(
+      new Tool("glob-search", "Search for symbols using glob pattern", schema),
+      (exchange, arguments) => {
+        try {
+          val query = arguments.get("query").asInstanceOf[String]
+          withPath(arguments) { path =>
+            queryEngine
+              .globSearch(query, Set.empty, path)
+              .map(result =>
+                new CallToolResult(
+                  createContent(result.map(_.show).mkString("\n")),
+                  false,
+                )
+              )
+          }.toMono
+        } catch {
+          case e: Exception =>
+            Mono.just(
+              new CallToolResult(createContent(s"Error: ${e.getMessage}"), true)
+            )
+        }
+      },
+    )
+  }
+
+  private def createTypedGlobSearchTool(): AsyncToolSpecification = {
+    val schema = """
+      {
+        "type": "object",
+        "properties": {
+          "query": {
+            "type": "string",
+            "description": "Substring of the symbol to search for"
+          },
+          "symbolType": {
+            "type": "array",
+            "items": {
+              "type": "string",
+              "enum": ["package", "class", "object", "function", "method", "trait"]
+            },
+            "description": "The type of symbol to search for"
+          },
+          "fileInFocus": {
+            "type": "string",
+            "description": "The current file in focus for context, if empty we will try to detect it"
+          }
+        },
+        "required": ["query", "symbolType"]
+      }
+    """
+    new AsyncToolSpecification(
+      new Tool(
+        "typed-glob-search",
+        "Search for symbols by type using glob pattern",
+        schema,
+      ),
+      (exchange, arguments) => {
+        try {
+          val query = arguments.get("query").asInstanceOf[String]
```

Comment:
Same, here and some cases below. Let's verify the type and not do asInstanceOf. I guess we can throw with our exception if this is not a string. Ideally we would have some kind of schema validator, but that might be an overkill for now.